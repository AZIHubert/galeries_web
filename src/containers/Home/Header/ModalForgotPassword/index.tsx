import { useFormik } from 'formik';
import * as React from 'react';

import Field from '#components/Field';

import { resetPasswordSchema } from '#helpers/schemas';

import ModalContainer from '#components/ModalContainer';
import GradientButton from '#components/GradientButton';

import { resendResetPassword } from '#helpers/api';

import {
  CancelButton,
  CancelButtonContainer,
} from './styles';

interface ModalForgotPasswordI {
  setCurrentEmail: React.Dispatch<React.SetStateAction<string>>;
  setError: React.Dispatch<React.SetStateAction<string>>;
  setForgotPassword: React.Dispatch<React.SetStateAction<boolean>>;
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
  switchModal: () => void;
  loading: boolean;
}

const initialValues = {
  email: '',
};

const ModalForgotPassword = ({
  setCurrentEmail,
  setError,
  setForgotPassword,
  setLoading,
  switchModal,
  loading,
}: ModalForgotPasswordI) => {
  const formik = useFormik({
    initialValues,
    onSubmit: async (value, { setFieldError }) => {
      if (!loading) {
        setLoading(true);
        try {
          await resendResetPassword(value);
          setCurrentEmail(value.email);
          switchModal();
        } catch (err) {
          if (err.response) {
            if (err.status === 500) {
              setError('Something went wrong. Please try again');
            } else {
              const { errors } = err.response.data;
              if (typeof errors === 'object') {
                Object.keys(errors).map((error) => setFieldError(error, errors[error]));
              } else {
                setError(errors);
              }
            }
          } else {
            setError('Something went wrong. Please try again');
          }
        }
        setLoading(false);
      }
    },
    validateOnChange: false,
    validateOnBlur: true,
    validationSchema: resetPasswordSchema,
  });
  return (
    <ModalContainer
      testId="modalForgotPassword"
      title='Enter your email to reset your password'
    >
      <form onSubmit={formik.handleSubmit}>
        <Field
          disabled={loading}
          id='email'
          error={formik.errors.email}
          errorTestId='emailError'
          fieldTestId='emailField'
          label='email'
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          touched={formik.touched.email}
          value={formik.values.email}
        />
        <GradientButton
          testId='submitButton'
          disabled={loading}
          marginBottom={20}
          marginTop={20}
          type='submit'
          title='Reset'
        />
      </form>
      <CancelButtonContainer>
        <CancelButton
          testId='cancelButton'
          onClick={() => {
            if (!loading) {
              setForgotPassword(false);
            }
          }}
        >
          Cancel
        </CancelButton>
      </CancelButtonContainer>
    </ModalContainer>
  );
};

export default ModalForgotPassword;
