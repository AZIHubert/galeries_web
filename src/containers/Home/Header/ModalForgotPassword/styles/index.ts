export { default as CancelButtonContainer } from './CancelButtonContainer';
export { default as CancelButton } from './CancelButton';
