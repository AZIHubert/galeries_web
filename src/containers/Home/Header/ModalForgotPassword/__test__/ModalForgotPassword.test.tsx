import * as React from 'react';
import renderer from 'react-test-renderer';
import {
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';

import {
  EMAIL_FIELD,
  REQUIRED,
} from '#helpers/formErrors';

import ModalForgotPassword from '../index';

const Container = () => {
  const [loading, setLoading] = React.useState(false);
  const setForgotPassword = () => {};
  const switchModal = () => {};
  const setCurrentEmail = () => {};
  return (
    <ModalForgotPassword
      setCurrentEmail={setCurrentEmail}
      setForgotPassword={setForgotPassword}
      setLoading={setLoading}
      switchModal={switchModal}
      loading={loading}
    />
  );
};

describe('ModalLogin', () => {
  const mockedSetForgotPassword = jest.fn;
  const mockedSetLoading = jest.fn;
  const mockedSetCurrentEmail = jest.fn;
  const mockedSwitchModal = jest.fn;
  let emailField: HTMLElement;
  let submitButton: HTMLElement;
  afterEach(cleanup);
  beforeEach(() => {
    const { getByTestId } = render(<Container />);
    emailField = getByTestId('emailField');
    submitButton = getByTestId('submitButton');
  });
  it('renders without crashing', () => {
    const tree = renderer.create(
      <ModalForgotPassword
        setCurrentEmail={mockedSetCurrentEmail}
        setForgotPassword={mockedSetForgotPassword}
        setLoading={mockedSetLoading}
        switchModal={mockedSwitchModal}
        loading={false}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should not loading if errors', async () => {
    fireEvent.click(submitButton);
    const emailError = await screen.findByTestId('emailError');
    expect(emailError).toHaveTextContent(REQUIRED);
    expect(emailField).not.toBeDisabled();
    expect(submitButton).toHaveTextContent('reset');
    expect(submitButton).not.toBeDisabled();
  });
  it('should loading if no error', async () => {
    fireEvent.change(emailField, { target: { value: 'user@email.com' } });
    fireEvent.click(submitButton);
    await screen.findByText('loading');
    expect(emailField).toBeDisabled();
    expect(submitButton).toBeDisabled();
  });
  describe('should return an error if', () => {
    describe('email', () => {
      it('is empty', async () => {
        fireEvent.blur(emailField);
        const emailError = await screen.findByTestId('emailError');
        expect(emailError).not.toBeNull();
        expect(emailError).toHaveTextContent(REQUIRED);
      });
      it('is not an email', async () => {
        fireEvent.change(emailField, { target: { value: 'notAnEmail' } });
        fireEvent.blur(emailField);
        const emailError = await screen.findByTestId('emailError');
        expect(emailError).not.toBeNull();
        expect(emailError).toHaveTextContent(EMAIL_FIELD);
      });
    });
  });
});
