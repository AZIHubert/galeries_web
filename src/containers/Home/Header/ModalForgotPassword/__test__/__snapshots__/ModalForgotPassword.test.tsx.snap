// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalLogin renders without crashing 1`] = `
<div
  className="sc-jSgupP dUUXuc"
  data-testid="modalForgotPassword"
>
  <div
    className="sc-gKsewC YbtFW"
  >
    <h3
      className="sc-iBPRYJ cuQqXR"
    >
      Enter your email to reset your password
    </h3>
    <form
      onSubmit={[Function]}
    >
      <div
        className="sc-bdfBwQ jVgFdX"
      >
        <label
          className="sc-hKgILt fWofNq"
          htmlFor="email"
        >
          <p>
            email
          </p>
        </label>
        <input
          className="sc-dlfnbm jEuCWi"
          data-testid="emailField"
          disabled={false}
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <button
        className="sc-fubCfw bOjZyX"
        data-testid="submitButton"
        disabled={false}
        type="submit"
      >
        Reset
      </button>
    </form>
    <div
      className="sc-pFZIQ isKpBa"
    >
      <button
        className="sc-jrAGrp gZzais"
        data-testid="cancelButton"
        onClick={[Function]}
      >
        Cancel
      </button>
    </div>
  </div>
  <div
    className="sc-eCssSg ksXhAV"
  />
</div>
`;
